### PR TITLE
feat(python): add snapshot restore support

### LINKED ISSUE #1256: feat(python): add snapshot/restore support
## Context

Node bindings (`crates/bashkit-js/`) support `snapshot()`, `restoreSnapshot()`, and `Bash.fromSnapshot()` for serializing and restoring interpreter state. Python bindings (`crates/bashkit-python/`) have no equivalent. This blocks stateful workflow patterns (checkpoint/resume, caching warm interpreters, etc.).

This is part of the Python ↔ Node binding parity effort (Phase 1 — API Parity).

## What to implement

Add three methods to `PyBash` and `PyBashTool` in `crates/bashkit-python/src/lib.rs`:

```python
# Instance methods
bash.snapshot() -> bytes        # serialize interpreter state to bytes
bash.restore_snapshot(data: bytes) -> None  # restore state from bytes

# Static factory
Bash.from_snapshot(data: bytes, **kwargs) -> Bash  # create new instance from snapshot
```

### Node reference implementation

In `crates/bashkit-js/src/lib.rs` (lines ~648-686), Node implements:
- `snapshot()` — calls the core `Interpreter::snapshot()`, returns `Uint8Array`
- `restore_snapshot(data)` — calls `Interpreter::restore_snapshot()`
- `Bash.fromSnapshot(data, options)` — constructs a new `Bash` with options, then restores

Python should follow the same pattern using PyO3's `&PyBytes` / `Vec<u8>`.

## Acceptance criteria

- [ ] `Bash.snapshot()` returns `bytes` containing serialized interpreter state
- [ ] `Bash.restore_snapshot(data)` restores state from `bytes` produced by `snapshot()`
- [ ] `Bash.from_snapshot(data, **kwargs)` creates a new `Bash` instance from snapshot bytes, accepting the same kwargs as `Bash()` constructor (`username`, `hostname`, `max_commands`, etc.)
- [ ] `BashTool` also has `snapshot()`, `restore_snapshot()`, and `BashTool.from_snapshot()`
- [ ] Round-trip works: variables, files, and CWD survive `snapshot()` → `from_snapshot()`
- [ ] Snapshot from one instance can restore into a different instance
- [ ] `restore_snapshot()` after `reset()` works correctly
- [ ] Invalid/corrupted snapshot data raises `BashError` (not a panic)
- [ ] Empty snapshot (from fresh instance) round-trips correctly
- [ ] Type stubs updated in `_bashkit.pyi`
- [ ] Tests added covering all above scenarios
- [ ] `cargo clippy` and `ruff check` pass
- [ ] README updated with Snapshot/Restore section

### PR DESCRIPTION
**Pull request by oliverlambson**
Closes #1256

## Summary
- add `snapshot()`, `restore_snapshot()`, and `from_snapshot()` to the Python `Bash` and `BashTool` bindings
- preserve constructor kwargs when restoring into a fresh instance from snapshot bytes
- add Python tests, stubs, and docs for round-trip, reset/restore, empty snapshot, and invalid snapshot cases

## Testing
- `cargo test -p bashkit-python --no-run`
- `cargo fmt --check`
- `.uv-venv-bashkit/bin/python -m pytest crates/bashkit-python/tests/test_bashkit.py`
- `.uv-venv-bashkit/bin/ruff check crates/bashkit-python`
- `cargo clippy -p bashkit-python --all-targets --no-deps -- -D warnings`

## Notes
- full `cargo clippy -p bashkit-python --all-targets -- -D warnings` still hits pre-existing dead-code warnings in `crates/bashkit` unrelated to this change

### CODE REVIEW

File: crates/bashkit-python/src/lib.rs [hunk 1]
```diff
@@ -725,6 +725,43 @@ fn glob_via_bash(rt: &Arc<Runtime>, inner: &Arc<Mutex<Bash>>, pattern: String) -
...
+fn snapshot_live_bash(
+    py: Python<'_>,
+    rt: &Arc<Runtime>,
+    inner: &Arc<Mutex<Bash>>,
+) -> PyResult<Vec<u8>> {
+    let rt = rt.clone();
+    let inner = inner.clone();
+    py.detach(|| {
+        rt.block_on(async move {
+            let bash = inner.lock().await;
```
[the remark this answers is not in the document]
Comment (oliverlambson) [reply crates/bashkit-python/src/lib.rs:743]: the concern is real in theory, but it is not a PR-specific inconsistency: a lot of the existing sync Python helper paths already do py.detach + rt.block_on + inner.lock().await without the execute_sync() timeout guard

to me it feels like standardizing all sync helper methods is a separate PR/issue

File: crates/bashkit-python/src/lib.rs [hunk 2]
```diff
@@ -725,6 +725,43 @@ fn glob_via_bash(rt: &Arc<Runtime>, inner: &Arc<Mutex<Bash>>, pattern: String) -
...
+    py: Python<'_>,
+    rt: &Arc<Runtime>,
+    inner: &Arc<Mutex<Bash>>,
+    data: Vec<u8>,
+) -> PyResult<()> {
+    let rt = rt.clone();
+    let inner = inner.clone();
+    py.detach(|| {
+        rt.block_on(async move {
+            let mut bash = inner.lock().await;
```
[the remark this answers is not in the document]
Comment (oliverlambson) [reply crates/bashkit-python/src/lib.rs:759]: see https://github.com/everruns/bashkit/pull/1298#discussion_r3083163799

File: specs/013-python-package.md [hunk 3]
```diff
@@ -130,6 +130,10 @@ tool = BashTool(files={
     "/config/static.txt": "ready\n",
     "/config/generated.json": lambda: '{"ok": true}\n",
 })
+# Snapshot / restore state
+blob = tool.snapshot()
+restored = BashTool.from_snapshot(blob, username="user")
+restored.restore_snapshot(blob)
```
[the remark this answers is not in the document]
Comment (oliverlambson) [reply specs/013-python-package.md]: fixed in [510f34a](https://github.com/everruns/bashkit/pull/1298/commits/510f34a7844552861ae7a29c4a12d10bdbf40ee3)

File: crates/bashkit-python/tests/test_bashkit.py [hunk 4]
```diff
@@ -545,6 +602,44 @@ def test_reset():
...
+    snapshot = tool.snapshot()
+
+    tool.reset()
+    tool.restore_snapshot(snapshot)
+
+    assert tool.execute_sync("echo $KEEP").stdout.strip() == "1"
+    assert tool.execute_sync("pwd").stdout.strip() == "/workspace"
+    assert tool.execute_sync("cat /workspace/tool.txt").stdout.strip() == "saved"
+
+
```
[the remark this answers is not in the document]
Comment (oliverlambson) [reply crates/bashkit-python/tests/test_bashkit.py:642]: fixed in [510f34a](https://github.com/everruns/bashkit/pull/1298/commits/510f34a7844552861ae7a29c4a12d10bdbf40ee3)

**Comment by chaliy** [comment]
LGTM, 

@oliverlambson pls assign PR to me when you feel it ready for merge.

**Comment by oliverlambson** [comment]
@chaliy great!

I don't have permissions to assign PRs on this repo, but I'm happy for you to self-assign & merge 👍

**Comment by chaliy** [comment]
Thank you!
